DROOLS-3425 : Unexpected behavior of docks when multiple editors are opened

Rikkola opened the pull request:
https://issues.jboss.org/browse/DROOLS-3425

More PRs to come, just pushing this to see what tests fail.

https://github.com/kiegroup/kie-wb-common/pull/2503
https://github.com/kiegroup/drools-wb/pull/1082
https://github.com/kiegroup/jbpm-designer/pull/837
https://github.com/kiegroup/jbpm-wb/pull/1309
https://github.com/kiegroup/optaplanner-wb/pull/327

Rikkola commented:
Does draft PR fire the build?

manstis commented on kie-wb-common-widgets/kie-wb-metadata-widget/src/main/java/org/kie/workbench/common/widgets/metadata/client/KieEditor.java:
```diff
@@ -51,23 +57,42 @@
         extends BaseEditor<T, Metadata>
         implements KieEditorWrapperView.KieEditorWrapperPresenter {
 
+    //This implementation always permits closure as something went wrong loading the Editor's content
```
This class seems to have had some crazy re-ordering of elements.. any chance you could revert?

Rikkola replied:
@manstis done as well as the other changes you requested.

manstis commented:
@paulovmr I thought you'd want to review given the changes to Library.

@wmedvede I thought you'd want to review given the changes to Data Modeller.

manstis commented:
@karreiro I thought you'd want to review given the changes to DMN's Decision Navigator.

Rikkola commented:
Fixing the GWT compile issues and any unit test issues that this build picks up. Can't get the unit tests in kie-wb-common to pass, so I need to abuse the PR build for it.

Rikkola commented:
Due to this touching so many editors, I'll try to target the merge to next week so this does not break anything for the release.

Rikkola commented:
Fixed the issues ( I think ). Updated the PR to see if the build is fine with this. I might have to add some unit tests, I'll do that tomorrow. See you later.

Rikkola commented:
@wmedvede this should compile now.

Rikkola commented:
jenkins please retest this

yurloc commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp/deps-master.txt:
```diff
@@ -0,0 +1,497 @@
+[INFO] Scanning for projects...
```
Committed accidentally, I guess.

Rikkola replied:
Thanks. These indeed are accidentally in, used these to track unnecessary dependencies I might have introduced.

yurloc commented on kie-wb-common-dmn/kie-wb-common-dmn-project-client/src/main/java/org/kie/workbench/common/dmn/project/client/editor/DMNDiagramEditor.java:
```diff
@@ -196,9 +209,8 @@ public void onOpen() {
-        decisionNavigatorDock.close();
-        decisionNavigatorDock.resetContent();
         dataTypesPage.disableShortcuts();
+        super.onClose();
```
Please rename `superOnClose()` (line 313) to `superDoClose()` as its what it does and it's confusing here to see
```java
public void onClose() {
    superOnClose();
    ...
    super.onClose();
}
```

Rikkola replied:
@yurloc Good point onSomething() indicates the change happens by an event and doSomething() is used to indicate that what ever is done actually happens.

Rikkola commented:
@jomarko your marked as a suggested person to review and @kkufova is marked as a tester in the ticket. Probably not necessary for both to check, but I'll let you decide who is the best from QA or if you both want to take a look.

Rikkola commented:
@yurloc The changes you requested are done.

kkufova commented:
@Rikkola, I think it won't hurt if we both take a look as this is quite a big PR. Is it ready for review or are we still waiting for some changes?

jomarko commented on kie-wb-common-screens/kie-wb-common-project-imports-editor/kie-wb-common-project-imports-editor-client/src/main/java/org/kie/workbench/common/screens/projectimportsscreen/client/forms/ProjectImportsScreenPresenter.java:
```diff
@@ -178,6 +185,7 @@ public IsWidget asWidget() {
     @OnClose
```
May we add `@Override` please?

jomarko commented on kie-wb-common-screens/kie-wb-common-workbench/kie-wb-common-workbench-client/src/main/java/org/kie/workbench/common/workbench/client/docks/AuthoringWorkbenchDocks.java:
```diff
@@ -82,6 +85,10 @@ public void initialize() {
                                                               handler.init(() -> setActiveHandler(handler)));
     }
 
+    public boolean isSetup() {
```
Could we add `@Override` please?

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-project-client/src/main/java/org/kie/workbench/common/stunner/bpmn/project/client/editor/BPMNDiagramEditor.java:
```diff
@@ -145,6 +145,7 @@ public void onOpen() {
     @OnClose
     public void onClose() {
```
Could we please add `@Override`?

jomarko commented on kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/docks/DefaultEditorDock.java:
```diff
@@ -0,0 +1,56 @@
+
+import org.uberfire.mvp.PlaceRequest;
+
+public class DefaultEditorDock {
```
seems to me as new class, would you mind add some test? if it is just renamed one, ignore my request. Just one more question, is there any specific reason to use `Instance<EditorDock> dock;` as package private instead of regular private?

Rikkola replied:
You points are on point and I have no defence. I'll make the field private and add the test.

jomarko commented:
### Manual review feedback
The only thing I spotted is that switching between multiple opened assets doesn't respect the open/closed state of the docks, however id doesn't work neither without changes introduced here, so I can approve once Toni have a look on my comments to code. Thank you @Rikkola!

Rikkola commented:
@jomarko yes that is why I had to add this action to when the screen is visible and focused and when this state is lost. The onClose only activates when the screen is actually closed, not when you change from one screen to another.

Rikkola commented:
@kkufova All the changes should be in now. Just adding annotations that should not change behaviour.

Rikkola commented on kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/docks/DataModellerDocksHandler.java:
```diff
@@ -90,11 +93,11 @@ public boolean isGraphicMode(DataModelerContext context) {
                                         perspectiveIdentifier).withSize(450).withLabel(constants.DocksAdvancedTitle()));
 
-            if (authorizationManager.authorize(WorkbenchFeatures.PLANNER_AVAILABLE,
+            if (authorizationManager.authorize(PLANNER_AVAILABLE,
```
Yes did this to avoid a dependendency on the WorkbenchFeatures module just for the one field, but maybe I can move WorkbenchFeatures to a shared location

Rikkola commented:
All updates should be in.

wmedvede commented:
hi @Rikkola , I'll take a look at this by the end of today, sorry for the delay.
Thanks!

wmedvede commented:
Jenkins execute full downstream build

Rikkola commented:
Added few Override annotations for onClose() methods that I previously missed.

Rikkola commented:
jenkins please retest this

Looks like the sources were out of sync during the previous build.

Rikkola commented:
Jenkins execute full downstream build

Rikkola commented:
I noticed the same thing with the right side docks and not sure what causes it.

@kiegroup/gatekeepers The bundle is now all green and approved, ready to be merged.
